### PR TITLE
Allow multiple database connections in a single Harlequin session

### LINKED ISSUE #53: Support switching active Postgres database during a Harlequin session
## Problem

  The Postgres adapter currently connects to a single database for the lifetime of a Harlequin session. Although the catalog can show multiple databases from the Postgres server, editor queries continue to run against the database selected at startup.

  If a user wants to query another database on the same server, they need to quit Harlequin and relaunch it with a different `dbname`.

  This is awkward when working with Postgres instances that expose multiple related databases, because the catalog makes those databases visible but there is no way to make one of them the active query target from inside the app.

  ## Desired Behavior

  Users should be able to select a database from the catalog and make it the active database for future editor queries without restarting Harlequin.

  A minimal version of this would:

  - Show all connectable databases in the catalog, as today.
  - Add a `Connect to Database` interaction to database catalog items.
  - When selected, switch the adapter’s active query connection to that database.
  - Keep catalog expansion database-specific, so expanding database `A` shows objects from `A` even if editor queries are currently targeting database `B`.
  - Preserve the existing active connection if switching fails.
  - Avoid switching while a manual transaction is active, since silently abandoning or moving a transaction would be surprising.

  ## Why This Matters

  This would make multi-database Postgres servers much easier to use from Harlequin. Users could browse several databases in the catalog, switch query context from the UI, and continue working in one session instead of restarting the app for each database.

  ## Potential Future Direction

  A more complete future design could allow each editor tab or query buffer to maintain its own database context. That would let one tab run against `db_a` while another tab runs against `db_b`.

  The initial implementation can use a single global active database for the session, because per-tab database context likely needs support from Harlequin core to pass buffer or tab identity through query execution.

### PR DESCRIPTION
**Pull request by jplarose**
## Summary

  Resolves #53.

  This PR adds multi-database support to the Postgres adapter so a Harlequin session can browse and switch between databases on the same Postgres server.

  The main change is that the adapter now keeps a lazy cache of connection pools by database. PostgreSQL connections are database-scoped, so catalog queries for another database need a real connection to
  that database rather than reusing the currently active connection. Creating pools lazily lets us avoid connecting to every database up front while still reusing pools after a database has been visited.

  ## Changes

  - Added database switching through a new `Connect to Database` catalog interaction.
  - Added one connection pool per visited database, with the active query connection swapped when switching databases.
  - Updated catalog loading so schemas, relations, columns, and materialized views are fetched from the database being inspected.
  - Preserved libpq/environment/default database behavior when no database name is explicitly provided.
  - Closed all cached pools when the adapter connection is closed.
  - Pinned the local Docker Postgres image to `postgres:17`.

  ## Edge Cases Covered

  Tests were added for:

  - Reusing an existing pool when switching back to a previously visited database.
  - Preserving the current active connection when target pool creation fails.
  - Preserving the current active connection when target connection checkout fails.
  - Returning a borrowed target connection if transaction-mode sync fails.
  - Avoiding a half-switched state if returning the previous connection fails.
  - Rejecting database switches while a manual transaction is active.
  - Respecting default/environment-selected databases instead of forcing `postgres`.
  - Verifying catalog schema lookups borrow from the requested database without changing the active database.
  - Verifying the new catalog interaction calls `switch_database` and notifies the user.

### CODE REVIEW

**Comment by tconbeer** [comment]
Thanks. I think this is more of a Harlequin feature than a postgres-adapter feature. Hot-swapping multiple database connections, even across different db adapters, is something that will come to Harlequin eventually, when I'm not so busy at my day job.

**Comment by jplarose** [comment]
> Thanks. I think this is more of a Harlequin feature than a postgres-adapter feature. Hot-swapping multiple database connections, even across different db adapters, is something that will come to Harlequin eventually, when I'm not so busy at my day job.

That makes sense! Could be an update to the implementation contract for adapters in `HarlequinConnection` where the DB specific implementation still lives in the adapter. Or were you thinking at a higher level than that? Potentially a combination of Harlequin controlling a pool of different adapters connections and the adapters controlling a pool  of their local connections?

**Comment by tconbeer** [comment]
Hey @jplarose I'm revisiting this idea, and while my comment above about multiple adapters/connections still stands, I think that Postgres is a little different from other database engines that let you jump between logical databases more seamlessly... Harlequin already works fine for those (e.g., DuckDB, MySQL). So I think I've changed my mind and I'd like to get this merged into harlequin-postgres, if you're willing to help me a little more on it.

I don't have time right now for a full code review, but will try to do that soon. Before I do, with the benefit of more time, do you still feel good about this implementation? Anything that needs to get updated before we ship it?

**Comment by jplarose** [comment]
> Hey @jplarose I'm revisiting this idea, and while my comment above about multiple adapters/connections still stands, I think that Postgres is a little different from other database engines that let you jump between logical databases more seamlessly... Harlequin already works fine for those (e.g., DuckDB, MySQL). So I think I've changed my mind and I'd like to get this merged into harlequin-postgres, if you're willing to help me a little more on it.
> 
> I don't have time right now for a full code review, but will try to do that soon. Before I do, with the benefit of more time, do you still feel good about this implementation? Anything that needs to get updated before we ship it?

Hi @tconbeer ! I think it is relatively good to go as it stands! I've been using my branch locally since I stood it up and it has worked exactly how I would like it to. The only odd thing that happened to me the other night was that my DB Server had to reboot while I had the connection open in Harlequin and it wasn't able to recover by "switching databases" as I had hoped it would, but I think that is relatively expected in the event the entire connection is dropped due to a reboot on that side. 

Let me know what else I could do to help if you'd like to see anything else on it before you feel good to ship it!

File: src/harlequin_postgres/adapter.py [hunk 1]
```diff
@@ -219,12 +294,10 @@ def _get_databases(self) -> list[tuple[str]]:
                 ;"""
             )
             results: list[tuple[str]] = cur.fetchall()
-        self.pool.putconn(conn)
         return results
 
     def _get_schemas(self, dbname: str) -> list[tuple[str]]:
-        conn: Connection = self.pool.getconn()
-        with conn.cursor() as cur:
+        with self._borrow_conn(dbname) as conn, conn.cursor() as cur:
```
Comment (tconbeer) [inline src/harlequin_postgres/adapter.py]: I think this is motivated by the limitation in `_get_mvs`, but for all of these other ones, we're just querying `information_schema`, which should yield the same results across all databases... It would be faster/more efficient to just use `self._borrow_conn()` for most of these, rather than the db-specific pools/connections.

I'm worried that this will effectively create connection pools for every db on the server in order to load the catalog, which I don't think we want.

**Review by tconbeer — CHANGES_REQUESTED** [review]
This looks really great -- just one change re: connections used for fetching the catalog, and one nit on exception handling.

Could you also please add an item to CHANGELOG.md?

File: src/harlequin_postgres/interactions.py [hunk 2]
```diff
@@ -77,6 +77,24 @@ def _drop_database() -> None:
...
+    if item.connection is None:
+        return
+    try:
+        item.connection.switch_database(item.label)
+    except HarlequinQueryError:
+        driver.notify(f"Could not connect to database {item.label}", severity="error")
+        raise
+    except Exception:
+        driver.notify(f"Could not connect to database {item.label}", severity="error")
+        raise
```
Comment (tconbeer) [inline src/harlequin_postgres/interactions.py]: `switch_database` only raises `HarlequinConnectionError`, not `HarlequinQueryError`:

```suggestion
    except HarlequinConnectionError:
        driver.notify(f"Could not connect to database {item.label}", severity="error")
        raise
```

**Comment by jplarose** [comment]
@tconbeer Addressed the review items and found a couple areas that still required the dbname to be passed for connection specific areas in Postgres. Made those connections temporary and short lived via `_borrow_catalog_conn` and validated by adding a couple tests and manually testing it against my server. 

Also updated the changelog and the incorrect exception type! If there is anything else that caught your eye let me know and I can take another look at it!

**Comment by jplarose** [comment]
Hello @tconbeer ! No problem if you have other priorities for the time being, just wanted to ping you on this to see if you had the chance to review my newer changes and if there was anything else I could take a look at to help make sure it is ready!
